src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.Async.cs

```diff
@@ -0,0 +1,254 @@
+        bool result = TryReadBlockCore(blockContents, bytesRead, out Zip64EndOfCentralDirectoryLocator? zip64EOCDLocator);
+        return (result, zip64EOCDLocator);
+    }
+
+    public static ValueTask WriteBlockAsync(Stream stream, long zip64EOCDRecordStart, CancellationToken cancellationToken)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        byte[] blockContents = new byte[TotalSize];
+        WriteBlockCore(blockContents, zip64EOCDRecordStart);
+        return stream.WriteAsync(blockContents, cancellationToken);
+    }
+}
+
+internal sealed partial class Zip64EndOfCentralDirectoryRecord
+{
+    public static async Task<(bool, Zip64EndOfCentralDirectoryRecord?)> TryReadBlock(Stream stream, CancellationToken cancellationToken)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        byte[] blockContents = new byte[BlockConstantSectionSize];
+        int bytesRead = await stream.ReadAtLeastAsync(blockContents, blockContents.Length, throwOnEndOfStream: false, cancellationToken).ConfigureAwait(false);
+        bool result = TryReadBlockCore(blockContents, bytesRead, out Zip64EndOfCentralDirectoryRecord? zip64EOCDRecord);
+        return (result, zip64EOCDRecord);
+    }
+
+    public static ValueTask WriteBlockAsync(Stream stream, long numberOfEntries, long startOfCentralDirectory, long sizeOfCentralDirectory, CancellationToken cancellationToken)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        byte[] blockContents = new byte[BlockConstantSectionSize];
+        WriteBlockCore(blockContents, numberOfEntries, startOfCentralDirectory, sizeOfCentralDirectory);
+        // write Zip 64 EOCD record
+        return stream.WriteAsync(blockContents, cancellationToken);
+    }
+}
+
+internal readonly partial struct ZipLocalFileHeader
+{
+    public static async Task<(List<ZipGenericExtraField>, byte[] trailingData)> GetExtraFieldsAsync(Stream stream, CancellationToken cancellationToken)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        // assumes that TrySkipBlock has already been called, so we don't have to validate twice
+
+        byte[] fixedHeaderBuffer = new byte[FieldLengths.FilenameLength + FieldLengths.ExtraFieldLength];
+        GetExtraFieldsInitialize(stream, out int relativeFilenameLengthLocation, out int relativeExtraFieldLengthLocation);
+        await stream.ReadExactlyAsync(fixedHeaderBuffer, cancellationToken).ConfigureAwait(false);
+
+        GetExtraFieldsCore(fixedHeaderBuffer, relativeFilenameLengthLocation, relativeExtraFieldLengthLocation, out ushort filenameLength, out ushort extraFieldLength);
+
+        byte[]? arrayPoolBuffer = extraFieldLength > StackAllocationThreshold ? ArrayPool<byte>.Shared.Rent(extraFieldLength) : null;
+        Memory<byte> extraFieldBuffer = extraFieldLength <= StackAllocationThreshold ? new byte[extraFieldLength] : arrayPoolBuffer.AsMemory(0, extraFieldLength);
```

Comment:
This logic won't make much sense in an async method - we can't allocate this buffer on the stack. We can use an ArrayPool unconditionally. This has knock-on effects with the redundant null check in the `finally` block, and might mean we no longer need to share the `StackAllocationThreshold` constant.

```suggestion
        byte[] arrayPoolBuffer = ArrayPool<byte>.Shared.Rent(extraFieldLength);
        Memory<byte> extraFieldBuffer = arrayPoolBuffer.AsMemory(0, extraFieldLength);
```

The same point also applies to ZipCentralDirectoryFileHeader.TryReadBlockAsync.

Reply:
Cool, applied. Made the const local again for the sole method that uses it.